---
title: "Assignment 6 RNAseqPart 2"
output: 
  html_document: 
    keep_md: yes
---

__Name:__ Zun Wang
__Student ID:__ 915109847

Insert your answers below.  Knit this file and submit the .Rmd and .html when you are done. 
```{r}
library(tidyverse)
library(goseq)
library(rtracklayer)
library(GenomicRanges)
library(Biostrings)
```

```{r}
DEgene.trt <- read_csv("../output/DEgenes.trt.csv")
head(DEgene.trt)
colnames(DEgene.trt)[1] <- "GeneID"
head(DEgene.trt)
```

__Exercise 1__:  
__a.__ Import the gene descriptions that you downloaded; pay attention to the "col_names" argument.  What is appropriate here?  Use one of the `join()` functions (which one?) to add gene descriptions for the genes found to be regulated by the DP treatment.  Output a table of the top 10 genes that includes the output from edgeR and the descriptions.
```{r}
descrip <- read_table("../input/FileS9.txt",col_names = "General") 
descrip <- descrip %>% separate(General,c("GeneID","Description"),"\t") #need to add column name here
```
```{r}
DEgene.trt <- inner_join(DEgene.trt,descrip)
```
```{r}
head(DEgene.trt,n=10)
```



__b.__ Repeat this for  genes with a genotype x trt interaction.
```{r}
DEgene.interaction <- read_csv("../output/DEgenes.interaction.csv")
colnames(DEgene.interaction)[1] <- "GeneID"
DEgene.interaction <- inner_join(DEgene.interaction,descrip)
head(DEgene.interaction,n=10)
```

__Exercise 2__:  
```{r}
go.terms <- read_tsv("../input/FileS11.txt",col_names=FALSE)
head(go.terms)
colnames(go.terms) <- c("GeneID","GO")
head(go.terms)

expressed.genes <- read_tsv("../input/internode_expressed_genes.txt")
head(expressed.genes)
names(expressed.genes) <- "GeneID"

gene.lengths <- read_tsv("../input/Brapa_CDS_lengths.txt")
head(gene.lengths)

#we need to reduce the gene.length data to only contain entries for those genes in our expressed.genes set.  We also need this as a vector
gene.lengths.vector <- gene.lengths$Length[gene.lengths$GeneID %in% expressed.genes$GeneID]
names(gene.lengths.vector) <- gene.lengths$GeneID[gene.lengths$GeneID %in% expressed.genes$GeneID]
head(gene.lengths.vector)

#Do the reverse to make sure everything matches up (it seems that we don't have length info for some genes?)
expressed.genes.match <- expressed.genes[expressed.genes$GeneID %in% names(gene.lengths.vector),]

```
```{r}
go.list <- strsplit(go.terms$GO,split=",")
names(go.list) <- go.terms$GeneID
head(go.list)
```
```{r}
DE.interaction <- expressed.genes.match$GeneID %in% DEgene.interaction$GeneID
    #for each gene in expressed gene, return FALSE if it is not in DEgene.trt and TRUE if it is.
names(DE.interaction) <- expressed.genes.match$GeneID
head(DE.interaction)
DE.interaction <- as.numeric(DE.interaction) #convert to 0s and 1s
head(DE.interaction)
sum(DE.interaction)
```
```{r}
#determines if there is bias due to gene length.  The plot shows the relationship.
nullp.result <- nullp(DEgenes = DE.interaction,bias.data = gene.lengths.vector)

#calculate p-values for each GO term
rownames(nullp.result) <- names(gene.lengths.vector) #because of a bug in nullp()
GO.out <- goseq(pwf = nullp.result,gene2cat = go.list,test.cats=("GO:BP"))

#list over-represented GO terms (p < 0.05)
GO.out[GO.out$over_represented_pvalue < 0.05,]
```
```{r}
write.table(GO.out[GO.out$over_represented_pvalue < 0.05,1:2],row.names=FALSE,file="../output/GO_terms.txt", quote = FALSE,col.names = FALSE)
```

__a:__ In REVIGO display a "TreeMap" of the BP GO terms.  Was our hypothesis that cell wall genes are enriched in the genotype X treatment gene set correct?  You DO NOT need to include the treemap in your answer.

Yes, the terms related to cell wall are overly represented in the selected GO terms, although not the biggest part but have its space in the BP GO treemap.

__b:__ Display a "TreeMap" of the CC GO terms.  There are four general categories shown, some with sub-categories.  What are the two general categories with the largest number of sub categories?  How might any of the four general categories relate to differences in plant growth?  (Note that the apoplast is the area outside of the plasma membrane, including the cell wall.  You DO NOT need to include the treemap in your answer.

The largest one is ER membrane, it might be related to membrane protein recycle rate. The second largest is cell wall, which might be related to plant cell wall organization. Others are cell surface and apoplast, which may related to cell signaling.

**SDR** -0.5 How do these relate to growth specifically?

__Exercise_3: __Flanking sequences could be defined in a variety of ways.  We want `flank()` to take sequences upstream, not downstream of the mRNAs.  Also, remember that genes can be transcribed from the "+" (upper) or "-" (lower) DNA strand.  Thus we want `flank()` to take coding strand into consideration and take sequences that are upstream of the 5' end of the gene.  Examine the mRNA and promoter ranges for the first couple of genes to determine if `flank()` has done what we want.  Explain how you know.

```{r}
gff <- import.gff("../../Assignment_5_Wang.Zun/input/Brapa_reference/Brapa_gene_v1.5.gff")
gff
```
```{r}
mRNAranges <- gff[gff$type=="mRNA",c("type", "ID")]
mRNAranges
```
```{r}
mRNAranges <- mRNAranges[str_detect(seqnames(mRNAranges), "Scaffold", negate = TRUE), ]
mRNAranges
```
```{r}
promoterRanges <- flank(mRNAranges, 1500)
promoterRanges
```

For + strands, flank takes the regions in numbers 1500 bases before the mRNA regions to be the promoter regions, and for - strands, flank takes the regions in numbers 1500 bases after the mRNA regions so that the promoters are counted from the 5' end before the mRNA, instead of only minus the number in regions.



__Exercise 4__  
```{r}
Brapaseq <- readDNAStringSet("../../Assignment_5_Wang.Zun/input/Brapa_reference/BrapaV1.5_chrom_only.fa")
```
```{r}
names(Brapaseq)
```
```{r}
names(Brapaseq) <- str_remove(names(Brapaseq), " \\[.*")
names(Brapaseq)
```
```{r}
promoters <- Brapaseq[promoterRanges]
names(promoters) <- promoterRanges$ID
promoters
```
```{r}
promoters <- DNAStringSet(gsub("N","-",promoters))

promoters
```

```{r}
motifs <- read.delim("../input/element_name_and_motif_IUPACsupp.txt",header=FALSE,as.is=TRUE)
head(motifs)
motifsV <- as.character(motifs[,2])
names(motifsV) <- motifs[,1]
motifsSS <- DNAStringSet(motifsV)
motifsSS
```
```{r}
DEgene.interaction.match <- DEgene.interaction$GeneID[DEgene.interaction$GeneID %in% names(promoters)]

expressed.genes.match <- expressed.genes$GeneID[expressed.genes$GeneID %in% names(promoters)]

#subset promoter files
universe.promoters <- promoters[expressed.genes.match]
target.promoters <- promoters[DEgene.interaction.match]
```

```{r}
#create a function to summarize the results and test for significance
motifEnrichment <- function(target.promoters,universe.promoters,all.counts=F,motifs=motifsSS) {
  
  #use vcountPDict to count the occurrences of each motif in each promoter
  target.counts <- vcountPDict(motifs,target.promoters,fixed=F) + 
    vcountPDict(motifsSS,reverseComplement(target.promoters),fixed=F)
  universe.counts <- vcountPDict(motifs,universe.promoters,fixed=F) + 
    vcountPDict(motifsSS,reverseComplement(universe.promoters),fixed=F)
  
  if (all.counts) { 
    #count all occurrences of a motif instead of the number of promoters that it occurs in
    target.counts.sum <- apply(target.counts,1,sum)
    universe.counts.sum <- apply(universe.counts,1,sum)
  } else {
    target.counts.sum <- apply(ifelse(target.counts > 0,1,0),1,sum)
    universe.counts.sum <- apply(ifelse(universe.counts > 0 , 1, 0),1,sum)
  }
  n.motifs <- length(target.counts.sum)
  results <- vector(mode="numeric",length=n.motifs)
  for (i in 1:n.motifs) {
    if (all.counts) { #the contigency tables are different depending on whether we are looking at promoters or overall occurrences
      #test if ratio of occurrences to promoters is the same in the target and the universe
      m <- matrix(c(
        target.counts.sum[i],                       #number of occurrences within target
        dim(target.counts)[2],                      #number of promoters in target
        universe.counts.sum[i],                  #number of occurrences within universe
        dim(universe.counts)[2]                  #number of promoters in universe
      ),ncol=2)
    } else { #looking at promoters with and without hits
      m <- matrix(c(
        target.counts.sum[i],                        #number of promoters in target with hit
        dim(target.counts)[2]-target.counts.sum[i],            #number of promoters in target with no hit
        universe.counts.sum[i],                   #number of promoters in universe with hit
        dim(universe.counts)[2]-universe.counts.sum[i]   #number of promoters in universe with no hit
      ),ncol=2)
    } #else
    results[i] <- fisher.test(m,alternative="greater")$p.value
  } #for loop
  results.table <- data.frame(
    motif=names(motifs),
    universe.percent = round(universe.counts.sum/dim(universe.counts)[2],3)*100,
    target.percent = round(target.counts.sum/dim(target.counts)[2],3)*100,
    p.value =  results)
  results.table <- results.table[order(results.table$p.value),]
  results.table
}
```
```{r}
motif.results <- motifEnrichment(target.promoters,universe.promoters)
head(motif.results)

```

__a.__ How many motifs are enriched at P < 0.05? 
```{r}
nrow(subset(motif.results,p.value<0.05))
```

__b.__ What is the identity of the most significantly over-enriched promoter?  
  SORLREP3

__c.__ What percentage of genes in the "Universe" have this motif?  What percentage in our target set?

in universe: 14.4
in target: 27.6

__d.__ You can find information on the motifs [here](http://arabidopsis.med.ohio-state.edu/AtcisDB/bindingsites.html).  Given the experimental treatments used in this experiment, do you think that the most enriched motif represents a biologically meaningful result?  Discuss why or why not.

yes, because it is about light regulation, and it corresponds perfectly with  the nature of sun/shadow treatment. We can intepret it as due to the light change, the plant regulate correspondently.

