---
title: "Assignment 4, Part II"
output: 
  html_document:
    keep_md: true
    df_print: paged
---
__Name:__ Zun Wang

__Student ID:__ 915109847

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, max.print=100)
```

## Assignment 4: Rice SNPs and GWAS

Remember to include the relevant code from the lab page in this file so that this file will knit.

**Exercise 0:** Be sure that your final knitted file has good formatting.  Make sure that you are using informative variable names.---
title: "Rice GWAS"
layout: lab
tags:
- R
- SNPs
hidden: true
---

```{r}
library(tidyverse)
load("../output/data_from_SNP_lab.Rdata")
```
```{r}
pheno.geno.pca.pop <- left_join(geno.pca.pop, data.pheno, by=c("ID" = "NSFTVID"))

#get rid of spaces in the phenotype names with "make.names()"
colnames(pheno.geno.pca.pop) <- make.names(colnames(pheno.geno.pca.pop))

head(pheno.geno.pca.pop)

dim(pheno.geno.pca.pop)
```

```{r}
library(ggplot2)
pheno.geno.pca.pop %>%
  ggplot(aes(x=`Seed.length.width.ratio`)) +
  geom_histogram()
```
```{r}
pl <- ggplot(data=pheno.geno.pca.pop, aes(x=Seed.length.width.ratio)) #create the basic plot object
pl <- pl + geom_histogram(binwidth=.5) #tell R that we want a histogram, with binwidth of .5
pl <- pl + facet_wrap(facets= ~ Region, ncol=3) # a separate plot ("facet") for each region, arranged in 3 columns
pl <- pl + ggtitle("Amylose Content") #add a title
pl #display the plot
```

**Exercise 7:** What Trait did you choose?

* Plot your chosen trait data
  * as a **single histogram** for all of the data
  * as **separate histograms** for each of the 4 population assignments made by fastStructure
  * as a **boxplot** separated by population assignments made by fastStructure. 
* Based on these histograms do you think that your trait varies by population? 
* **optional** Try using the "violin" geom.  
  
*Hint:* you will need to use a different binwidth than I used with Seed.length.width (or don't specify it at all and let R choose the default).

*Hint:* the relevant column name for population is "assignedPop".  
```{r}
pheno.geno.pca.pop %>%
  ggplot(aes(x=`Blast.resistance`)) +
  geom_histogram()
```

```{r}
bl <- ggplot(data=pheno.geno.pca.pop, aes(x=`Blast.resistance`)) #create the basic plot object
bl <- bl + geom_histogram() #tell R that we want a histogram, with binwidth of .5
bl <- bl + facet_wrap(facets= ~ assignedPop, ncol=3) # a separate plot ("facet") for each region, arranged in 3 columns
bl <- bl + ggtitle("Blast.resistance") #add a title
bl #display the plot
```
```{r}
bl2 <- ggplot(data=pheno.geno.pca.pop, aes(x=`Blast.resistance`)) #create the basic plot object
bl2 <- bl2 + geom_boxplot() #tell R that we want a histogram, with binwidth of .5
bl2 <- bl2 + facet_wrap(facets= ~ assignedPop, ncol=3) # a separate plot ("facet") for each region, arranged in 3 columns
bl2 <- bl2 + ggtitle("Blast.resistance") #add a title
bl2 #display the plot
```
```{r}
#The difference is big and is potentially significant.
```

**Exercise 8:** 

* Obtain the mean of your trait for each of the four fastStructure populations.
* Perform an ANOVA for your trait to test if it varies significantly by fastStructure population.  Show your code, the ANOVA output, and provide an interpretation.
* Discuss: Do your results suggest a problem for GWAS?
```{r}
sem <- function(x, na.rm=TRUE) {
  if(na.rm) x <- na.omit(x)
  sd(x)/sqrt(length(x)-1)
}

pheno.geno.pca.pop %>% group_by(assignedPop) %>% 
  summarize(mean.bl=mean(Blast.resistance,na.rm=T),
            sem.bl=sem(Blast.resistance)
            ) %>%
  arrange(desc(mean.bl))

aov2 <- aov(Blast.resistance ~ assignedPop,data=pheno.geno.pca.pop) #1-way ANOVA for Amylose.content by Region
summary(aov2)
#The blast resistance difference among populations are significant, so blast resistance is different among each population, and thus is problem for GWAS.
```

**Exercise 9:** Complete the code below to generate a GM object like the one above.  Note that "chr" and "pos" should be numeric.  _Hint:_ Look at help for the `separate` command.  You will need to specify the `into`, `sep`, `convert` and `remove` arguments.
```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("multtest")
```
```{r}
install.packages("EMMREML")
devtools::install_github("jiabowang/GAPIT3",force=TRUE)
devtools::install_github("YaoZhou89/BLINK")

```
```{r}
data.geno <- read_csv("../input/Rice_44K_genotypes.csv.gz",
                      na=c("NA","00"))  %>%
  rename(ID=X1) %>% 
  dplyr::select(-`6_17160794_1`)
```
```{r}
gtToNumeric <- function(x) {
  x <- as.numeric(as.factor(x))-1 # convert to numeric set base genotype to 0
  x[is.na(x)] <- mean(x, na.rm=TRUE) # fill in missing data. THIS SHOULD BE DONE BETTER!
  if(max(x, na.rm = TRUE)==1) 
    return(x*2) # set alternate homozygous to 2 if necessary
  else return(x) 
}

#Now apply it to all gt columns
#Base R function "apply" is much faster than using tidyverse functions when you have this many columns
GD <- cbind(data.geno[,1], apply(data.geno[-1], 2, gtToNumeric))

#take a quick look
head(GD[,1:10])
```
```{r, eval=TRUE}
# GAPIT is one of the rare cases where tibbles and data frames are not interchangeable
# So be sure to used "data.frame()" rather than "tibble()"
GM <- data.frame(SNP=colnames(GD)[-1]) %>%
  separate(SNP, into = c("chr","pos"), sep = "_", remove = TRUE, convert = FALSE ) 
GM <- cbind(SNP = colnames(GD)[-1],GM)
head(GM)
```

Because GAPIT requires a data frame, not a tibble, we convert after selecting the phenotype

Fill in the trait you chose in the code below and run the code.  What trait did you pick?


```{r, eval=FALSE}
Y <- pheno.geno.pca.pop %>%
  dplyr::select(ID, Blast.resistance) %>%  # remove the "<" and ">"
  as.data.frame()
```
```{r}
library(GAPIT3)
```



#### No structure correction

__I am going to leave the GAPIT code in the template__.  Note that I have the chunk option set to eval=FALSE.  This ensures that the code will not run when you knit.__  You will need to run it before knitting.

```{r, eval=FALSE}
if(!dir.exists("../output/GAPIT_GLM_no_Q")) dir.create("../output/GAPIT_GLM_no_Q")
setwd("../output/GAPIT_GLM_no_Q")

gapout <- GAPIT(
  Y=Y,
  GD=GD,
  GM=GM,
  model="GLM"
  )
```


#### Q matrix

```{r, eval=FALSE}
if(!dir.exists("../output/GAPIT_GLM_Q")) dir.create("../output/GAPIT_GLM_Q")
setwd("../output/GAPIT_GLM_Q")

Q <- pheno.geno.pca.pop %>%
  dplyr::select(ID, pop2, pop3, pop4) %>% # leave one pop out...it is used as the base level.
  as.data.frame()

gapout <- GAPIT(
  Y=Y,
  GD=GD,
  GM=GM,
  CV=Q,
  model="GLM")
```

#### PCA matrix

```{r eval=FALSE}
if(!dir.exists("../output/GAPIT_GLM_PCA")) dir.create("../output/GAPIT_GLM_PCA")
setwd("../output/GAPIT_GLM_PCA")

QPC <- PCs %>% as.data.frame()

system.time(gapout <- GAPIT(
  Y=Y,
  GD=GD,
  GM=GM,
  CV=QPC,
  model="GLM"
  ))
```

**Exercise 10:** Compare the Q-Q and Genome-wise Manhattan plots of the no correction, Q matrix, and PCA matrix runs.  (You will find the PDFs in the respective output folders).  Did the corrections make a difference?  If so, which one worked better?  How did this effect the number of "significant" SNPs in the Manhattan plot?  (In the Manhattan plot the horizontal line represents the significance threshold.  If you don't see a line, nothing was significant).

**Include your Q-Q and Genome-wise Manhattan plots in your Rmd file**.  You will need to convert them to png or jpeg, see below.
```{r}
#The correction, espectially PCA makes a difference. PCA makes more significant SNPs in the Manhattan plot.
```
![Manhattannoc](/home/ubuntu/LAB3/Assignment_4_Wang.Zun/output/pic/Manhattanno.png)
![ManhattanQ](/home/ubuntu/LAB3/Assignment_4_Wang.Zun/output/pic/ManhattanQ.png)
![ManhattanPCA](/home/ubuntu/LAB3/Assignment_4_Wang.Zun/output/pic/ManhattanPCA.png)
![qqnoc](/home/ubuntu/LAB3/Assignment_4_Wang.Zun/output/pic/QQ-plotno.png)
![qqQ](/home/ubuntu/LAB3/Assignment_4_Wang.Zun/output/pic/QQplotQ.png)
![qqPCA](/home/ubuntu/LAB3/Assignment_4_Wang.Zun/output/pic/QQplotPCA.png)

```{r}
```

#### K Matrix

```{r, eval=FALSE}
if(!dir.exists("../output/GAPIT_CMLM_no_Q")) dir.create("../output/GAPIT_CMLM_no_Q")
setwd("../output/GAPIT_CMLM_no_Q")

gapout <- GAPIT(
  Y=Y,
  GD=GD,
  GM=GM,
  group.from = 1,
  group.to = 413,
  model="CMLM"
  )
```


**Exercise 11:** Compare the Q-Q plot from the K matrix run to the best one from above.  Is the K matrix more or less effective at controlling for population structure? **Include the Q-Q and Genome-wise Manhattan plot in your Rmd file**.
```{r}
#It is very effective and better than others.
```
#### CMLM with Q and K

![qqc](/home/ubuntu/LAB3/Assignment_4_Wang.Zun/output/pic/2.png)
![manc](/home/ubuntu/LAB3/Assignment_4_Wang.Zun/output/pic/1.png)


We can combine K and either Q or PCA matrices.  Choose either "Q" or "QPC" based on whichever gave better results above and modify the code below.

```{r, eval=FALSE}
if(!dir.exists("../output/GAPIT_CMLM")) dir.create("../output/GAPIT_CMLM")
setwd("../output/GAPIT_CMLM")

gapout <- GAPIT(
  Y=Y,
  GD=GD,
  GM=GM,
  CV=QPC, ## either Q or QPC
  group.from=1,
  group.to=413,
  model="CMLM"
  )
```
![qqnew](/home/ubuntu/LAB3/Assignment_4_Wang.Zun/output/pic/3.png)
![mannew](/home/ubuntu/LAB3/Assignment_4_Wang.Zun/output/pic/4.png)
**Exercise 12:**
Does this do any better than the best GLM model with Q or QPC? Comment on both the QQ and Manhattan plots. **Include the Q-Q and Genome-wise Manhattan plot in your Rmd file**.
```{r}
#The pic is above, and it works even better than QPC alone. The dots follow the expectation line more and there are more significant SNPs.
```

#### Blink

Finally let's try the mystery method, "Blink"

```{r, eval=FALSE}
if(!dir.exists("../output/GAPIT_Blink")) dir.create("../output/GAPIT_Blink")
setwd("../output/GAPIT_Blink")

gapout <- GAPIT(
  Y=Y,
  GD=GD,
  GM=GM,
  model="Blink"
  )
```

**Exercise 13:** How does Blink compare to the other methods?  Again, comment on both the QQ and Manhattan plots. **Include the Q-Q and Genome-wise Manhattan plot in your Rmd file**.
```{r}
#This one looks horrible, the dots go away so far and there are few significant SNPs.
```
![qqblink](/home/ubuntu/LAB3/Assignment_4_Wang.Zun/output/pic/5.png)
![manblink](/home/ubuntu/LAB3/Assignment_4_Wang.Zun/output/pic/6.png)

**Exercise 14:** Load in the SNP list from whichever method best controlled population structure.   The file ends with `GWAS.Results.csv` and is in the relevant output folder. Use `head()` to show the top 10 SNPs.
```{r}
result <- read_csv("/home/ubuntu/LAB3/Assignment_4_Wang.Zun/output/GAPIT_CMLM/GAPIT._CMLM.Blast.resistance.GWAS.Results.csv")
head(result)
```

**Exercise 15:** Look for genes close to your SNP at the [rice genome browser](http://rice.plantbiology.msu.edu/cgi-bin/gbrowse/rice/).  Pick a significant SNP from your analysis and enter its chromosome and position in the search box.  The browser wants you to enter a start and stop position, so for example, you should enter "Chr3:30449857..30449857" and then choose "show 20kb" from the pulldown menu on the right hand side.  Report the SNP you chose and the three closest genes.  These are candidate genes for determining the phenotype of your trait of interest in the rice population.  Briefly discuss these genes as possible candidates for the GWAS peak.  __Include a Screenshot of the genome browser in your answer__
```{r}
#It can be one of the highest one in the 5 column, since I chose the first one to run.
```
![screenshots](/home/ubuntu/LAB3/Assignment_4_Wang.Zun/output/pic/s.png)
