---
title: "Assignment 4, Part I"
output: 
  html_document:
    keep_md: true
    df_print: paged
---
__Name:__ Zun Wang

__Student ID:__ 915109847

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, max.print=100)
```

## Assignment 4: Rice SNPs and GWAS

Remember to include the relevant code from the lab page in this file so that this file will knit.
```{r}
library(tidyverse)
```
**Exercise 0:** Be sure that your final knitted file has good formatting.  Make sure that you are using informative variable names.
```{r}
data.geno <- read_csv("../input/Rice_44K_genotypes.csv.gz",
                      na=c("NA","00"))
```
```{r}
data.geno <- data.geno %>% select(-`6_17160794_1`)
```
```{r}
head(data.geno[,1:10]) #first six rows of first 10 columns
summary(data.geno[,1:10]) #summarizes the first 10 columns
```
```{r}
data.geno <- data.geno %>% rename(ID=X1)

head(data.geno[,1:10])
```
**Exercise 1:** Create a data subset that contains a random sample of 10000 SNPs from the full data set.  Place the smaller data set in an object called `data.geno.10000`.  __Very important: you want to keep the first column, the one with the variety IDs, and you want it to be the first column in `data.geno.10000`.  AND You do not want it to show up randomly later on in the data set.__  Think about how to achieve this.

```{r}
?append
data.geno.10000 = data.geno[,append(1, sample(2:ncol(data.geno),10000))]
data.geno.10000
```
```{r}
dim(data.geno.10000)
colnames(data.geno.10000) %>% str_which("ID")
```
**Exercise 2:** plot the variance explained by the first 10 rows
```{r}
geno.numeric <- data.geno.10000[,-1] %>% # -1 to remove the first column, with names.
  lapply(factor) %>% # convert characters to "factors", where each category is internally represented as a number
  as.data.frame() %>% # reformat
  data.matrix() #  convert to numeric

head(geno.numeric[,1:10])
```
```{r}
geno.numeric.fill <-
  apply(geno.numeric, 2, function(x) {
    x[is.na(x)] <- mean(x, na.rm=T)
    x})
geno.pca <- prcomp(geno.numeric.fill, 
            rank.=10) # We really only need the first few, this tells prcomp to only return the first 10
str(geno.pca)
head(geno.pca$x)[,1:5]
head(geno.pca$rotation)[,1:5]
head(geno.pca$sdev)
pcvar <- geno.pca$sdev^2 # square std dev to get variance
pcvar.pct <- tibble(pctvar=pcvar/sum(pcvar) * 100,
                    PC=1:length(pcvar))
```
```{r}
plotdata <- data.frame(
  name=pcvar.pct$PC[1:10],
  value=pcvar.pct$pctvar[1:10]
)

ggplot(plotdata, aes(x=name, y=value)) + 
  geom_bar(stat = "identity")

```
```{r}
PCs <- as_tibble(geno.pca$x) %>% # the principal components
  mutate(ID=data.geno.10000$ID) %>%
  select(ID, everything())
head(PCs)
```
**Exercise 3:** Make 2 scatter plots, the first of PC1 vs PC2, and second PC2 vs PC3. Is there any evidence for populations structure (different sub populations)?  If so, how many sub populations do you think the MDS plot reveals?  What do you make of the individuals that are between the major groups?
```{r}
ggplot(PCs, aes(x=PC1, y=PC2)) +
  geom_point()

ggplot(PCs, aes(x=PC2, y=PC3)) +
  geom_point()

#There are evidences of subpopulations, and there are 4 subpopulation in total. The individuals between can be the crossed ones between the subgroups.
```
**Exercise 4:** 
* Use the `read_csv()` `head()` and `summary()` functions that you learned earlier to import and look at this file.  Import the file into an object called "data.pheno".
* Use a `join` function to merge the PC  genotype data (in the object `PCs`) with the phenotype data into a new object called "data.pheno.pca". Use summary and head to look at the new object and make sure that it is as you expect.  
* It (`data.pheno.pca`) should have 413 rows and 49 columns.
* Include your code in the .Rmd

```{r}
data.pheno <- read_csv("../input/RiceDiversity.44K.MSU6.Phenotypes.csv",
                      na=c("NA","00"))
head(data.pheno[,1:10]) 
summary(data.pheno[,1:10])

data.pheno.pca <- inner_join(PCs, data.pheno, by=c("ID"="NSFTVID"))

```
```{r}
dim(data.pheno.pca)
```
**Exercise 5:** Prepare three different PCA plots to explore if subgroups vary by 1) Amylose content; 2) Pericarp color; 3) Region.  That is make a scatter plot of PC1 vs PC2 and color the points by the above characteristics.  Do any of these seem to be associated with the different population groups? Briefly discuss.  (optionally repeat the plots plotting PC2 vs PC3)
```{r}
#Amylose
ggplot(data.pheno.pca, aes(x=PC1, y=PC2, color=`Amylose content`)) +
  geom_point()
# Pericapr color
ggplot(data.pheno.pca, aes(x=PC1, y=PC2, color=`Pericarp color`)) +
  geom_point()
# Region
ggplot(data.pheno.pca, aes(x=PC1, y=PC2, color=`Region`)) +
  geom_point()
#The plots show that the two subgroups seperate with distinctive characteristics, with Asian group related to higher amylose and darker pericarp color compared to America/Europe group.
```
`
**Exercise 6:** First, use a join function to combine the PCA data (in object `PCs`) with the population assignments (in `fs_results`) and place the result in `geno.pca.pop`  Then re plot the PCA data, but include the population assignment in an informative way.  How do the populations assignments relate to the PCA plot?
```{r}
data.geno.10000.fs <- matrix("",nrow=nrow(data.geno.10000)*2,ncol=ncol(data.geno.10000)-1+6)

for (i in 1:nrow(data.geno.10000)) {
  data.geno.10000.fs[(i-1)*2+1,1:6] <- data.geno.10000[[i,1]]
  data.geno.10000.fs[(i-1)*2+2,1:6] <- data.geno.10000[[i,1]]
  data.geno.10000.fs[(i-1)*2+1,-1:-6] <- substr(data.geno.10000[i,-1],1,1)
  data.geno.10000.fs[(i-1)*2+2,-1:-6] <- substr(data.geno.10000[i,-1],2,2)
}

data.geno.10000.fs[is.na(data.geno.10000.fs)] <- -9 # fastStructure's code for missing data

dim(data.geno.10000.fs)

#take a look
data.geno.10000.fs[1:10,1:10]

write.table(data.geno.10000.fs,file="../output/rice.data.fastStructure.input.str", col.names = FALSE, row.names = FALSE, quote = FALSE)
```
```{r}
fam <- tibble(
  FID=data.geno.10000$ID,
  IID=data.geno.10000$ID,
  PID=0,
  MID=0,
  Sex=0,
  Ptype=-9)

head(fam)
```
```{r}
bim <- data.geno.10000.fs[,-1:-6]

colnames(bim) <- colnames(data.geno.10000)[-1]

bim[bim=="-9"] <- NA

bim <- apply(bim,2,function(x) unique(na.omit(x))) 

bim[,1:5]
```
```{r}
bim <- t(bim) %>%  # t transposes the matrix
  as_tibble() %>%
  mutate(SNP_ID=colnames(bim), cM=0) 

head(bim)
```
```{r}
bim <- bim %>% 
  separate(SNP_ID,into = c("chromosome","position"),sep="_",remove=FALSE) %>% # create a column for chromosome and position
  select(chromosome, SNP_ID, cM, position, allele1=V1, allele2=V2) # get columns in right order

head(bim)
```
```{r}
write.table(bim,file="../output/rice.data.fastStructure.input.bim",col.names = FALSE, row.names = FALSE, quote = FALSE)
```

```{r}
fs_results <- read_delim("../output/rice.fastStructure.out.4.meanQ", delim="  ", col_names = FALSE, col_types = 'nnnn')
head(fs_results)
```


```{r}
fs_results <- fs_results %>% 
  mutate(ID=data.geno.10000$ID) %>% 
  select(ID, pop1=X1, pop2=X2, pop3=X3, pop4=X4)
head(fs_results)

fs_results$assignedPop <- apply(fs_results[,-1], 1, which.max)
head(fs_results)
table(fs_results$assignedPop)
fs_results$maxPr <- apply(fs_results[,2:5],1,max) 
fs_results <- fs_results %>% 
  arrange(assignedPop,desc(maxPr)) %>%
  mutate(plot.order=row_number())
fs_results_long <- fs_results %>% pivot_longer(pop1:pop4,names_to="population",               values_to="proportion")
head(fs_results_long)
fs_results_long %>%
  ggplot(aes(x=plot.order, y=proportion, color=population, fill=population)) + 
  geom_col()  +
  ylab("geno
       me proportion") + 
  scale_color_brewer(type="div") + scale_fill_brewer(type="div")
```


```{r}
fs_results <- fs_results %>% mutate(assignedPop=as.character(assignedPop))
geno.pca.pop <- inner_join(PCs, fs_results, by="ID")
ggplot(geno.pca.pop, aes(x=PC1, y=PC2, color=`assignedPop`)) +
  geom_point()
#The population shows in four assigned groups and align well with the previous PCA subgroups, indicating each subgroup have differenct amylose content and other attributes.
```

```{r}
save(data.pheno,geno.pca, PCs, geno.pca.pop,fs_results,file="../output/data_from_SNP_lab.Rdata")
```



